File: app/src/main/java/com/example/android/plaid/data/api/designernews/DesignerNewsService.java
/*
 * Copyright 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.android.plaid.data.api.designernews;

import com.example.android.plaid.BuildConfig;
import com.example.android.plaid.data.api.designernews.model.AccessToken;
import com.example.android.plaid.data.api.designernews.model.StoriesResponse;

import retrofit.Callback;
import retrofit.http.Body;
import retrofit.http.GET;
import retrofit.http.Header;
import retrofit.http.Headers;
import retrofit.http.POST;
import retrofit.http.Query;

/**
 * Modeling the Designer News API
 */
public interface DesignerNewsService {

    String ENDPOINT = "https://www.designernews.co/";

    String CLIENT_ID_QUERY = "?client_id=" + BuildConfig.DESIGNER_NEWS_CLIENT_ID;

    @GET("/api/v1/stories" + CLIENT_ID_QUERY)
    void getTopStories(@Query("page") Integer page,
                       Callback<StoriesResponse> callback);

    @GET("/api/v1/stories/recent" + CLIENT_ID_QUERY)
    void getRecentStories(@Query("page") Integer page,
                          Callback<StoriesResponse> callback);

    @GET("/api/v1/stories/search" + CLIENT_ID_QUERY)
    void search(@Query("query") String query, Callback<StoriesResponse> callback);

    @Headers({
            "grant_type: password",
            "client_id: " + BuildConfig.DESIGNER_NEWS_CLIENT_ID,
            "client_secret: " + BuildConfig.DESIGNER_NEWS_CLIENT_SECRET
    })
    @POST("/oauth/token")
    void login(@Header("username") String username,
               @Header("password") String password,
               @Body String ignored,  // can remove when retrofit releases this fix:
               // https://github.com/square/retrofit/commit/19ac1e2c4551448184ad66c4a0ec172e2741c2ee
               Callback<AccessToken> callback);

}

File: core/src/main/java/io/plaidapp/core/designernews/data/api/DesignerNewsService.kt
/*
 * Copyright 2018 Google LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.plaidapp.core.designernews.data.api

import io.plaidapp.core.data.api.EnvelopePayload
import io.plaidapp.core.designernews.data.login.model.AccessToken
import io.plaidapp.core.designernews.data.login.model.LoggedInUserResponse
import io.plaidapp.core.designernews.data.stories.model.Story
import io.plaidapp.core.designernews.data.stories.model.StoryResponse
import retrofit2.Call
import retrofit2.Response
import retrofit2.http.FieldMap
import retrofit2.http.FormUrlEncoded
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

/**
 * Models the Designer News API.
 *
 * v1 docs: https://github.com/layervault/dn_api
 * v2 docs: https://github.com/DesignerNews/dn_api_v2
 */
interface DesignerNewsService {

    @EnvelopePayload("stories")
    @GET("api/v2/stories")
    suspend fun getStories(@Query("page") page: Int?): Response<List<StoryResponse>>

    @EnvelopePayload("stories")
    @GET("api/v2/stories/{ids}")
    suspend fun getStories(@Path("ids") commaSeparatedIds: String): Response<List<StoryResponse>>

    @EnvelopePayload("users")
    @GET("api/v2/me")
    suspend fun getAuthedUser(): Response<List<LoggedInUserResponse>>

    @FormUrlEncoded
    @POST("oauth/token")
    suspend fun login(@FieldMap loginParams: Map<String, String>): Response<AccessToken>

    /**
     * Search Designer News by scraping website.
     * Returns a list of story IDs
     */
    @DesignerNewsSearch
    @GET("search?t=story")
    suspend fun search(
        @Query("q") query: String,
        @Query("p") page: Int?
    ): Response<List<String>>

    @EnvelopePayload("story")
    @POST("api/v2/stories/{id}/upvote")
    fun upvoteStory(@Path("id") storyId: Long): Call<Story>

    companion object {
        const val ENDPOINT = "https://www.designernews.co/"
    }
}
